Deprecation and Obsoletion experience

petemill opened the pull request:
## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request [as needed](https://github.com/brave/browser-laptop/wiki/Pull-request-process).
- [ ] Request a security/privacy review [as needed](https://github.com/brave/handbook/blob/master/development/security.md#how-to-request-a-security-review). (Ask a Brave employee to help if you cannot access this document.)

## Test Plan:
### 1
From fresh profile or any profile which has not run with this code yet:
 - Permanent banner is shown stating there are 10 days left
 - 1.a) With Brave-Core _not_ already installed, banner prompts to Download
 - 1.b) With Brave-Core _already_ installed, banner prompts to Launch it
 - 1.c) Browser is useable: address bar, click on links from other programs, etc

### 2
Using profile from 1), either adjust clock forward 10 days, or `deprecatedOn` time in profile state file back at least 10 days
 - Permanent banner text changes to state the browser is no longer useable
 - 2.a) Tabs and windows are no longer restored
 - 2.b) Browser is _not_ useable: address bar, click on links from other programs, etc
 - 2.c) NTP always displays obsoletion UI and prompts to either 2.c.i)launch or 2.c.ii) install brave-core

### 3
There is no option to update the browser, nor is there any message saying that there are no new updates to the browser.

## Reviewer Checklist:

- [ ] Request a security/privacy review [as needed](https://github.com/brave/handbook/blob/master/development/security.md#how-to-request-a-security-review) if one was not already requested.

Tests


- [ ] Adequate test coverage exists to prevent regressions
- [ ] Tests should be independent and work correctly when run individually or as a suite [ref](https://github.com/brave/browser-laptop/wiki/Code-Guidelines#test-dependencies)
- [ ] New files have MPL2 license header

petemill committed:
Revert to deprecated UI and automatically switch to obsolete UI in 10 days

petemill committed:
Permanent deprecation banner with 'days until obsoletion' countdown

petemill committed:
Obsoletion State unit tests

petemill commented:
Added unit tests because time and math...

bsclifton commented on app/renderer/components/main/main.js:
```diff
@@ -758,6 +791,30 @@ const styles = StyleSheet.create({
+
+  deprecationBanner_Message: {
+    color: globalStyles.color.commonTextColor,
+    fontSize: '15px'
```
this needs to have `cursor: default` and `user-select: none` added (it currently shows a text selection cursor and is selectable via double click)

bsclifton commented on js/about/newtab.js:
```diff
@@ -266,80 +275,187 @@ class NewTabPage extends React.Component {
       return <div className='deprecationNotice'>
+        <button
+          className='fa fa-close'
+          onClick={this.dismissNotice.bind(this)}
```
This warning shouldn't be dismissable, IMO (no `X` button)

bsclifton commented on app/renderer/components/main/main.js:
```diff
@@ -580,6 +585,33 @@ class Main extends React.Component {
   }
 
+  onDeprecationButtonClick = () => {
+    if (!this.props.braveCoreInstalled) {
```
should be `if (!this.props.isBraveCoreInstalled) {`

bsclifton commented on app/renderer/components/main/main.js:
```diff
@@ -580,6 +585,33 @@ class Main extends React.Component {
 
+  onDeprecationButtonClick = () => {
+    if (!this.props.braveCoreInstalled) {
+      appActions.maybeCreateTabRequested({
```
this was re-worked into `appActions.createTabRequested` (which is confusing, because the shields reference above was never updated! whoops)

petemill committed:
New Tab deprecation notice is no longer dismissable

petemill committed:
Obsoletion UX: New Tab page and Banner text

petemill commented:
Un-marking as WIP now since all features are now implemented. I still want to take another look at the code tomorrow as well as add test plans here and possibly a couple more unit tests and a way to manually test.

![image](https://user-images.githubusercontent.com/741836/55533650-d6afa900-5666-11e9-9933-a2f754865964.png)

![image](https://user-images.githubusercontent.com/741836/55533675-ed560000-5666-11e9-9308-8d2ff370242d.png)

bsclifton commented:
Fresh profile - I'm getting the fireworks ☹️ Which means I can't use the browser (during the 10 day period). Also, check for update happens (at least when running from source) and fails, showing a 2nd bar
<img width="1408" alt="Screen Shot 2019-04-04 at 1 08 34 PM" src="https://user-images.githubusercontent.com/4733304/55585188-f1236a00-56da-11e9-85b6-dfacbbfac485.png">

bsclifton commented on app/common/state/obsoletionStateHelper.js:
```diff
@@ -0,0 +1,50 @@
+}
+
+module.exports.getIsObsolete = function getIsObsolete (state) {
+  return true || getDaysUntilObsolete(state) === 0
```
I think this is meant to be the other way around 😄 as-is, it'll always return true

bsclifton committed:
Remove update bar (and menu items referencing it)

diracdeltas commented:
i doubt anyone will bother to do this, but just noting that someone can avoid Brave getting into the unusable state by setting their clock back

petemill committed:
After obsoletion date, do not open urls from cmd-line or OS

petemill commented:
Yeah @diracdeltas there are a couple ways to get around it (e.g. adjust the state file to put a deprecation date in the future, open the console and change window.location, build from source...), but the aim is to make it extremely obvious and pressured, not necessarily to be 100% secure AFAIK. If anyone has a different opinion then lmk.

petemill commented:
Test Plan added. I'm done unless I get any further feedback @bsclifton

bsclifton commented:
Couple of quick notes before reviewing 😄 

I did the following:
- archived master as `0.26.x` branch
- reset master to `0.25.x` 
- updated the base for this PR to be `master`

We'll want to make this version `0.27.x`. I'll review what we have here and then create a commit with that version

bsclifton committed:
0.27.0

bsclifton commented:
macOS and Linux builds available here:
https://github.com/brave/browser-laptop/releases

Didn't get a chance to fully test/review yet- also need to sort out issues with signing on Windows build servers

bsclifton reviewed (changes requested):
pricacy ☹️

srirambv commented:
GitHub shows insecure icon,when opening the rev version from `about:brave`
![image](https://user-images.githubusercontent.com/17010094/55683642-de45ab00-590f-11e9-9371-88f57424218d.png)
However opening support page from `Learn more...` shows secure connection for support.brave.com
![image](https://user-images.githubusercontent.com/17010094/55683660-2238b000-5910-11e9-8f21-7d27338d3bc6.png)

srirambv commented:
Context menu should probably be removed once the 10 days limit is reached. You can still continue to browse using the search functionality. Two options
1. Disable context menu
2. Remove `Search` option in context menu

I'd vote for option1 which makes it more hard to use the browser and forces to switch to bc
![muon](https://user-images.githubusercontent.com/17010094/55683735-3335f100-5911-11e9-9156-d1ee1cdaa492.gif)

srirambv commented:
Disable payment optin message if user has not opted into payments
![image](https://user-images.githubusercontent.com/17010094/55683769-65475300-5911-11e9-98e5-ee583b03d89a.png)

srirambv commented:
Setting system to an older date still give the browser to show X days left deprecation message and make it fully useable. Once the 10 days are completed irrespective of user manually setting system date to an older date, it should still show that browser has completed the 10 days grace period of usage from upgrade to `0.27.0` and make it useless.

bsclifton committed:
Don't show ledger notifications

bsclifton committed:
0.27.1

bsclifton committed:
s/pricacy/privacy

bsclifton commented:
Pushed some updates:
- fixed spelling error
- bumped version
- disabled ledger notifications completely

RE: the context menu- that is a good catch. But I think we're fine not fixing it. It's an edge case and would be a painfully inefficient way to use the browser in a daily driver context

petemill commented:
> Setting system to an older date still give the browser to show X days left deprecation message and make it fully useable. Once the 10 days are completed irrespective of user manually setting system date to an older date, it should still show that browser has completed the 10 days grace period of usage from upgrade to `0.27.0` and make it useless.

We’re not trying to make it securely impossible to browse any pages. Rather we are just disabling the obvious UX. After all, if someone really wanted to keep using this version, they could run from source, update the state JSON file with a different deprecatedOn date, or download an older release and disable connection to update server. There’s a line but I think we decided that the obvious functionality to disable is the address bar, tab restore on startup, and links from other apps. The other side of that line would be usage scenarios that are just too painful for someone to keep using this version as their main browser, like running with a clock that’s always in the past.

GeetaSarvadnya commented:
After 10 days the user can still access Bookmarks. 

![image](https://user-images.githubusercontent.com/38657976/55715343-29e07d80-5a12-11e9-8ca0-26d1bac0b671.png)

petemill committed:
When obsolete, disable opening bookmarks

This mainly prevents the bookmark bar actions, and stops short of disabling opening pages from the context menu of the bookmark page, to minimize risk of affecting normal webpage operations downloading the latest version of the browser

petemill commented:
Just pushed functionality which blocks bookmarks bar items from opening content, when obsolete. Stopped short of blocking context menu on bookmarks page as I did not want to add more risk to normal website operations for downloading the latest version.

bsclifton committed:
0.27.2

bsclifton reviewed (approved):
Beautiful! <3

bsclifton commented:
@GeetaSarvadnya yup- that is OK, IMO 😄Eventually sites in AC will drop off and that is a pretty painful way to experience the internet

The context menu is something @srirambv noted too - I think it's OK, as browsing via context menu is definitely a bad experience

Thanks for testing so quickly and great job being so thorough! 😄

LaurenWags commented:
Mentioned in channel, but will document here as well: 
```2.c) NTP always displays obsoletion UI and prompts to either 2.c.i)launch or 2.c.ii) install brave-core```
If user has 'A new tab shows' set to something other than 'Dashboard', the fireworks/obsoletion UI will not display after 10 days. We should probably force new tab page to show dashboard with warning message prior to 10 days.

LaurenWags commented:
mentioned in channel, will document here as well:
In the following scenario I'm not getting the "fireworks" NTP. Here are my steps:
1. Install 0.25.304, make it your default browser --> update to 0.27.2 (keep NTP as Dashboard)
2. Visit sites, etc as normal. Ignore warnings.
3. Close Brave, set computer clock ahead 10 days.
4. Click on a link from an app (such as Slack)
5. Muon launches and shows the link you clicked on.
Note, if you open muon, then click on a link, the link does not open - you see fireworks page (this is expected)

GeetaSarvadnya commented:
Verified both scenarios which @LaurenWags mentioned on version 0.27.3 both scenarios working fine on windows

Scenario1:
https://github.com/brave/browser-laptop/pull/15349#issuecomment-481860953

- A new tab dropdown is in a disabled state, the user doesn't have the option to set it different options.
- After setting the system date to 10 days ahead, launch brave is showing "fireworks" NTP

Scenario2:
https://github.com/brave/browser-laptop/pull/15349#issuecomment-481884322

- Upgraded muon from 0.25.304 to 0.27.3. A new tab dropdown is in a disabled state, the user doesn't have the option to set it different options.
- After setting system date to 10 days ahead, click on a link from slack app opened a "fireworks" NTP. clicked link did not open in 0.27.3

LaurenWags commented:
0.27.3 looks good on macOS
- Verified test plan from description
- Verified items from https://github.com/brave/browser-laptop/pull/15349#issuecomment-481280234, https://github.com/brave/browser-laptop/pull/15349#issuecomment-481860953, and https://github.com/brave/browser-laptop/pull/15349#issuecomment-481884322

btlechowski commented:
0.27.3 looks good on Windows 10 x64
- Verified test plan from description
- Verified items from https://github.com/brave/browser-laptop/pull/15349#issuecomment-481280234, https://github.com/brave/browser-laptop/pull/15349#issuecomment-481860953, and https://github.com/brave/browser-laptop/pull/15349#issuecomment-481884322

Minor issues found:
- After installing Brave core, the banner on top and banner on new tab do not change to `Launch the new Brave`. Browser restart fixes the issue.
- Browser laptop prompts to make the browser default. Doubt that we should do it when deprecating the browser.
- After 10 days passed, the `Launch the new Brave` is not visible on the viewport. the user has to scroll
![image](https://user-images.githubusercontent.com/34715963/55997687-a9dd4080-5cbb-11e9-9f1e-294ccc964d7e.png)

kjozwiak commented:
Verification **PASSED** on `macOS 10.14.4 x64` using the following build:

```
Brave: 0.27.3 
rev: 2df8cb101008dce51a22fe652dcf22b00c21f62e 
Muon: 8.1.8 
libchromiumcontent: 69.0.3497.100
```

Went through the following cases:

* Went through the test plan outlined under https://github.com/brave/browser-laptop/pull/15349#issue-266860983
* Went through the cases outlined under https://github.com/brave/browser-laptop/pull/15349#issuecomment-481280234
* Went through the case outlined under https://github.com/brave/browser-laptop/pull/15349#issuecomment-481860953 and ensured that `Dashboard` is selected as the default choice and cannot be changed once updated to `0.27.3 CR: 69.0.3497.100`
* Went through the cases outlined under https://github.com/brave/browser-laptop/pull/15349#issuecomment-481884322
* ensured that updated from `0.25.304 CR: 69.0.3497.100` --> `0.27.3 CR: 69.0.3497.100` worked without any issues.

kjozwiak commented:
@bsclifton @rebron looks like we're in good shape for the Monday release:

* @GeetaSarvadnya & @btlechowski went through Windows
* @LaurenWags and myself went through macOS
* @srirambv went through Linux

However, as mentioned under Slack, there's no way we'll be able to upgrade Linux as the repo and snaps have been converted to `b-c`.

@GeetaSarvadnya @LaurenWags @btlechowski @srirambv thanks for your all hard work and thoroughness!
